autotx/utils/LlamaClient.py

```diff
@@ -0,0 +1,81 @@
+from types import SimpleNamespace
+from typing import Any, Dict, Union, cast
+from autogen import ModelClient
+from llama_cpp import (
+    ChatCompletion,
+    ChatCompletionRequestMessage,
+    ChatCompletionRequestToolMessage,
+    ChatCompletionResponseMessage,
+    Completion,
+    CreateChatCompletionResponse,
+    Llama,
+)
+
+
+class LlamaClient(ModelClient):  # type: ignore
+    def __init__(self, _: dict[str, Any], **args: Any):
+        self.llm: Llama = args["llm"]
+
+    def create(self, params: Dict[str, Any]) -> SimpleNamespace:
+        sanitized_messages = self._sanitize_chat_completion_messages(
+            cast(list[ChatCompletionRequestMessage], params.get("messages"))
+        )
+        response = self.llm.create_chat_completion(
+            messages=sanitized_messages,
+            tools=params.get("tools"),
+            model=params.get("model"),
+        )
+
+        return SimpleNamespace(**{**response, "cost": "0"})  # type: ignore
+
+    def message_retrieval(
+        self, response: CreateChatCompletionResponse
+    ) -> list[ChatCompletionResponseMessage]:
+        choices = response.choices  # type: ignore
+        return [choice["message"] for choice in choices]
+
+    def cost(self, _: Union[ChatCompletion, Completion]) -> float:
+        return 0.0
+
+    def get_usage(self, _: Union[ChatCompletion, Completion]) -> dict[str, Any]:
+        return {
+            "prompt_tokens": 0,
+            "completion_tokens": 0,
+            "total_tokens": 0,
+            "cost": 0,
+            "model": "meetkai/functionary-small-v2.4-GGUF",
+        }
+
+    def _sanitize_chat_completion_messages(
```

Comment:
What does this do?

Reply:
when doing a tool call, the message from the chat completion doesn't return the name of the function w/the response - this is what the llamacpp bindings expect
```json
{
  tool_call_id": tool_call.id,
  "role": "tool",
  "name": function_name,
  "content": function_response,
}
```
so we're manually adding the `name` to that object to make things work as expected